crates/wasmtime/src/component/values.rs

```diff
@@ -0,0 +1,389 @@
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum Val {
+    /// Unit
+    Unit,
+    /// Boolean
+    Bool(bool),
+    /// Signed 8-bit integer
+    S8(i8),
+    /// Unsigned 8-bit integer
+    U8(u8),
+    /// Signed 16-bit integer
+    S16(i16),
+    /// Unsigned 16-bit integer
+    U16(u16),
+    /// Signed 32-bit integer
+    S32(i32),
+    /// Unsigned 32-bit integer
+    U32(u32),
+    /// Signed 64-bit integer
+    S64(i64),
+    /// Unsigned 64-bit integer
+    U64(u64),
+    /// 32-bit floating point value
+    Float32(u32),
+    /// 64-bit floating point value
+    Float64(u64),
+    /// 32-bit character
+    Char(char),
+    /// Character string
+    String(Box<str>),
+    /// List of values
+    List(List),
+    /// Record
+    Record(Record),
+    /// Tuple
+    Tuple(Tuple),
+    /// Variant
+    Variant(Variant),
+    /// Enum
+    Enum(Enum),
+    /// Union
+    Union(Union),
+    /// Option
+    Option(Option),
+    /// Expected
+    Expected(Expected),
+    /// Bit flags
+    Flags(Flags),
+}
+
+impl Val {
+    /// Retrieve the [`Type`] of this value.
+    pub fn ty(&self) -> Type {
+        match self {
+            Val::Unit => Type::Unit,
+            Val::Bool(_) => Type::Bool,
+            Val::S8(_) => Type::S8,
+            Val::U8(_) => Type::U8,
+            Val::S16(_) => Type::S16,
+            Val::U16(_) => Type::U16,
+            Val::S32(_) => Type::S32,
+            Val::U32(_) => Type::U32,
+            Val::S64(_) => Type::S64,
+            Val::U64(_) => Type::U64,
+            Val::Float32(_) => Type::Float32,
+            Val::Float64(_) => Type::Float64,
+            Val::Char(_) => Type::Char,
+            Val::String(_) => Type::String,
+            Val::List(List { ty, .. }) => Type::List(ty.clone()),
+            Val::Record(Record { ty, .. }) => Type::Record(ty.clone()),
+            Val::Tuple(Tuple { ty, .. }) => Type::Tuple(ty.clone()),
+            Val::Variant(Variant { ty, .. }) => Type::Variant(ty.clone()),
+            Val::Enum(Enum { ty, .. }) => Type::Enum(ty.clone()),
+            Val::Union(Union { ty, .. }) => Type::Union(ty.clone()),
+            Val::Option(Option { ty, .. }) => Type::Option(ty.clone()),
+            Val::Expected(Expected { ty, .. }) => Type::Expected(ty.clone()),
+            Val::Flags(Flags { ty, .. }) => Type::Flags(ty.clone()),
+        }
+    }
+
+    /// Serialize this value as core Wasm stack values.
+    pub(crate) fn lower<T>(
+        &self,
+        store: &mut StoreContextMut<T>,
+        options: &Options,
+        dst: &mut MaybeUninit<[ValRaw; MAX_STACK_PARAMS]>,
+        offset: usize,
+    ) -> Result<()> {
+        match self {
+            Val::Unit => (),
+            Val::Bool(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(if *value { 1 } else { 0 }));
+            }
+            Val::S8(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i32(*value as i32));
```

Comment:
Idiomatically we try to avoid `as` casts where possible since in isolation they could indicate that a lossy conversion is happening. Could the casts here be replaced with `i32::from(*value)`? That should signify that no data is lost.